From c034a346943a00adbdbc396a588db41001a6a045 Mon Sep 17 00:00:00 2001
From: Chris Tarazi <ctarazi@pivotal.io>
Date: Fri, 3 May 2019 10:54:05 -0700
Subject: [PATCH] Skip symlink for /proc if inside Docker and chroot

In our use case, we have a Docker container (because of our CI) that
runs `debootstrap` inside a chroot. We ran into issues with mounting
/proc in this environment. The error we were seeing was "Too many
symbolic links" because mount was detecting that /proc was a symlink to
itself.

This change skips the creation of the symlink if you are inside Docker
and a chroot to avoid the above error.

Co-authored-by: Joshua Aresty <joshua.aresty@emc.com>
Co-authored-by: Nicholas Calugar <ncalugar@pivotal.io>

Detect the presence of chroot with CHROOT_CMD

Co-authored-by: Joshua Aresty <joshua.aresty@emc.com>
Co-authored-by: Nicholas Calugar <ncalugar@pivotal.io>
---
 functions             | 4 ++++
 scripts/aequorea      | 2 +-
 scripts/debian-common | 2 +-
 scripts/edgy          | 2 +-
 scripts/feisty        | 2 +-
 scripts/gutsy         | 2 +-
 6 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/functions b/functions
index 6b936176b56b..8818ff11e556 100644
--- a/functions
+++ b/functions
@@ -1667,6 +1667,10 @@ on_exit () {
 	N_EXIT_THINGS=$(( $N_EXIT_THINGS + 1 ))
 }
 
+in_chroot() {
+	test -n "$CHROOT_CMD"
+}
+
 ############################################################## fakechroot tools
 
 install_fakechroot_tools () {
diff --git a/scripts/aequorea b/scripts/aequorea
index 8c689838cddf..1be96b028b0f 100644
--- a/scripts/aequorea
+++ b/scripts/aequorea
@@ -89,7 +89,7 @@ second_stage_install () {
 		baseprog="$(($baseprog + ${1:-1}))"
 	}
 
-	if doing_variant fakechroot || [ "$CONTAINER" = "docker" ]; then
+	if doing_variant fakechroot || ([ "$CONTAINER" = "docker" ] && ! in_chroot ); then
 		setup_proc_symlink
 	else
 		setup_proc
diff --git a/scripts/debian-common b/scripts/debian-common
index a2acd361fcb7..152ebf558daf 100644
--- a/scripts/debian-common
+++ b/scripts/debian-common
@@ -65,7 +65,7 @@ first_stage_install () {
 
 	setup_devices
 
-	if doing_variant fakechroot || [ "$CONTAINER" = "docker" ]; then
+	if doing_variant fakechroot || ([ "$CONTAINER" = "docker" ] && ! in_chroot ); then
 		setup_proc_symlink
 	fi
 }
diff --git a/scripts/edgy b/scripts/edgy
index d3b1ca124a51..f5cb412abc1c 100644
--- a/scripts/edgy
+++ b/scripts/edgy
@@ -87,7 +87,7 @@ second_stage_install () {
 	baseprog="$(($baseprog + ${1:-1}))"
     }
 
-    if doing_variant fakechroot || [ "$CONTAINER" = "docker" ]; then
+    if doing_variant fakechroot || ([ "$CONTAINER" = "docker" ] && ! in_chroot ); then
 	setup_proc_symlink
     else
 	setup_proc
diff --git a/scripts/feisty b/scripts/feisty
index f3b4031d36ac..e6a6ad896ad0 100644
--- a/scripts/feisty
+++ b/scripts/feisty
@@ -86,7 +86,7 @@ second_stage_install () {
 	baseprog="$(($baseprog + ${1:-1}))"
     }
 
-    if doing_variant fakechroot || [ "$CONTAINER" = "docker" ]; then
+    if doing_variant fakechroot || ([ "$CONTAINER" = "docker" ] && ! in_chroot ); then
 	setup_proc_symlink
     else
 	setup_proc
diff --git a/scripts/gutsy b/scripts/gutsy
index 82c59c50a892..533a3c271f16 100644
--- a/scripts/gutsy
+++ b/scripts/gutsy
@@ -92,7 +92,7 @@ first_stage_install () {
 
 	setup_devices
 
-        if doing_variant fakechroot || [ "$CONTAINER" = "docker" ]; then
+	if doing_variant fakechroot || ([ "$CONTAINER" = "docker" ] && ! in_chroot ); then
 		setup_proc_symlink
 	fi
 }
-- 
2.24.0

